api/src/main/java/org/pmiops/workbench/billing/FreeTierBillingService.java

```diff
@@ -85,35 +84,73 @@ public double getWorkspaceFreeTierBillingUsage(DbWorkspace dbWorkspace) {
    * passing thresholds or exceeding limits
    */
   public void checkFreeTierBillingUsage() {
-    // retrieve the costs stored in the DB from the last time this was run
-    final Map<DbUser, Double> previousUserCosts = workspaceFreeTierUsageDao.getUserCostMap();
+    // Get existing DB cost information.
+    final List<WorkspaceCostView> dbCost = workspaceDao.getWorkspaceCostViews();
 
-    // retrieve current workspace costs from BigQuery and store in the DB
-    final Map<DbWorkspace, Double> workspaceCosts = getFreeTierWorkspaceCostsFromBQ();
-    workspaceCosts.forEach(workspaceFreeTierUsageDao::updateCost);
+    // Grab BigQuery cost data.
+    final Map<String, Long> workspaceByProject =
+        dbCost.stream()
+            .collect(
+                Collectors.toMap(
+                    WorkspaceCostView::getGoogleProject, WorkspaceCostView::getWorkspaceId));
+    final Map<Long, Double> liveCostByWorkspace =
+        getFreeTierWorkspaceCostsFromBQ(workspaceByProject);
+
+    // Generate before/after workspace costs, find delta, apply updates
+    final Map<Long, Double> dbCostByWorkspace =
+        dbCost.stream()
+            .collect(
+                Collectors.toMap(
+                    WorkspaceCostView::getWorkspaceId,
+                    v -> Optional.ofNullable(v.getFreeTierCost()).orElse(0.0)));
+    liveCostByWorkspace.keySet().stream()
+        .filter(id -> compareCosts(dbCostByWorkspace.get(id), liveCostByWorkspace.get(id)) != 0)
+        .map(id -> workspaceDao.findById(id).get())
```

Comment:
I believe this can be optimized further by getting all the filtered workspaces using findAllByWorkspaceIdIn instead of making one query per workspace.